### PR TITLE
SPM to Specimen Mapping

### PR DESCRIPTION
**Pull request by brick-green**
This PR maps SPM to Specimen and back to SPM according to the mapping inventory

## Linked Issues
- linked #12416

### CODE REVIEW

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+collection:
+  condition: $spm7 NOT_NULL || $spm8 NOT_NULL || $spm9 NOT_NULL || $spm12 NOT_NULL
+  expressionType: nested
+  generateList: true
+  vars:
+    spm7: STRING, SPM.7
+    spm8: STRING, SPM.8
+    spm9: STRING_ALL, SPM.9
+    spm12: STRING, SPM.12
+  expressionsMap:
```
Comment (mkalish) [inline prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml:77]: Do each of these need null checks for their individual values?

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+      valueOf: datatypes/CWE/CodeableConcept
+      expressionType: resource
+      specs: SPM.27
+
+collection:
+  condition: $spm7 NOT_NULL || $spm8 NOT_NULL || $spm9 NOT_NULL || $spm12 NOT_NULL
+  expressionType: nested
+  generateList: true
+  vars:
+    spm7: STRING, SPM.7
```
Comment (mkalish) [inline prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml:73]: Maybe?
```suggestion
    spm7: STRING_ALL, SPM.7
```

File: prime-router/metadata/fhir_transforms/senders/SimpleReport/simple-report-sender-transform.yml [hunk 3]
```diff
@@ -319,13 +319,42 @@ elements:
...
+  ## Code
+  - name: sr-specimen-bodySite-code-system-extension
+    resource: 'Bundle.entry.resource.ofType(Specimen)'
+    bundleProperty: '%resource.collection.bodySite.coding[0].extension("https://reportstream.cdc.gov/fhir/StructureDefinition/cwe-coding-system").value[x]'
+    value: [ '%resource.collection.bodySite.coding[0].system.getCodingSystemMapping()' ]
+
+  # Populating display based on text
+  - name: sr-specimen-bodySite-display
+    resource: 'Bundle.entry.resource.ofType(Specimen)'
+    condition: '%resource.collection.bodySite.exists() and %resource.type.coding[0].display.exists().not()'
```
Comment (mkalish) [inline prime-router/metadata/fhir_transforms/senders/SimpleReport/simple-report-sender-transform.yml:354]: Do the other ones modifying `collection.bodySite` need this condition?

**Review by mkalish — COMMENTED** [review]
Looking really good!  I'll do another pass tomorrow with a closer look to the inventory.

Only real comment is if there needs to be additional null checks in some of the HL7-> FHIR mappings

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml (hunk 2 above)
Comment (brick-green) [reply prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml]: ran some tests and it should be either `STRING_ALL` or no type. So I opted for no type

**Review by mkalish — APPROVED** [review]
Few tiny things to tweak, but looks good to me.  Confirmed that this matches the inventory

**Comment by mkalish** [comment]
@brick-green Looks like some failing tests

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml (hunk 1 above)
Comment (brick-green) [reply prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml:90]: Only the ones with repeats need null checks

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/SPMExtension.yml [hunk 4]
```diff
@@ -0,0 +1,237 @@
+url:
+  type: SYSTEM_URL
+  value: spm-specimen
+
+extension:
+  generateList: true
+  expressionType: nested
+  expressions:
+    - expressionType: nested
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/fhir_mapping/hl7/segments/SPM/SPMExtension.yml:9]: Done

File: prime-router/src/testIntegration/resources/datatests/mappinginventory/eip/EIP-to-Identifier-FillerAssignedIdentifier.fhir [hunk 5]
```diff
@@ -134,16 +134,19 @@
...
+      }, {
+        "extension" : [ {
+          "url" : "https://reportstream.cdc.gov/fhir/StructureDefinition/hl7-use",
+          "valueString" : "other-specimen-id"
+        } ]
       } ],
-      "subject" : {
-        "reference" : "Patient/1703276815504458000.89f166df-1bef-429f-94a8-cea55ebb1eb1"
-      }
+      "accessionIdentifier" : { }
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/src/testIntegration/resources/datatests/mappinginventory/eip/EIP-to-Identifier-FillerAssignedIdentifier.fhir]: condition added

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+      valueOf: datatypes/EIP/Identifier-FillerAssignedIdentifier
+      expressionType: resource
+      specs: SPM.2
+      vars:
+        spm22: SPM.2.2
+      constants:
+        hl7Use: "specimen-id-filler"
+        backupExtension: "placer-assigned-identifier"
+    - valueOf: datatypes/CX/Identifier
+      expressionType: resource
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml:35]: condition added

File: prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+      specs: SPM.6 *
+    type:
+      valueOf: datatypes/CWE/CodeableConcept
+      expressionType: resource
+      specs: SPM.27
+
+collection:
+  condition: $spm7 NOT_NULL || $spm8 NOT_NULL || $spm9 NOT_NULL || $spm12 NOT_NULL
+  expressionType: nested
+  generateList: true
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/fhir_mapping/hl7/segments/SPM/Specimen.yml:83]: If all those fields are blank, won't `collection` still be made?

File: prime-router/metadata/hl7_mapping/resources/Specimen/SPM.yml [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    resource: '%resource.collection.method'
+    condition: '%resource.exists()'
+    resourceIndex: cweIndex
+    constants:
+      cweField: '%{hl7SpecimenFieldPath}-7(%{cweIndex})'
+    schema: ../../datatypes/codeableConcept/CWE
+
+  - name: specimen-source-site
+    resource: '%resource.collection.bodySite'
+    condition: '%resource.exists()'
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/hl7_mapping/resources/Specimen/SPM.yml]: yep, good call. Removed

File: prime-router/metadata/hl7_mapping/resources/Specimen/SPM.yml [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+    schema: ../../datatypes/dateTime/DTMorDT
+
+  - name: specimen-observation-date-time-start
+    condition: '%context.collection.collected.exists() and %context.collection.collected is Period'
+    resource: '%resource.collection.collected.start'
+    constants:
+      dtmFieldPath: '%{hl7SpecimenFieldPath}-17-1'
+    schema: ../../datatypes/dateTime/DTMorDT
+
+  # Alternative to getting specimen collection time from DiagnosticReport instead
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/metadata/hl7_mapping/resources/Specimen/SPM.yml]: another good call, removed

File: prime-router/src/testIntegration/resources/datatests/mappinginventory/nk1/nk1.fhir
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/src/testIntegration/resources/datatests/mappinginventory/nk1/nk1.fhir:1]: nope, accidental adds

File: prime-router/src/testIntegration/kotlin/datatests/mappinginventory/spm/SPMtoSpecimen.kt [hunk 10]
```diff
@@ -0,0 +1,16 @@
+package gov.cdc.prime.router.datatests.mappinginventory.spm
+
+import gov.cdc.prime.router.datatests.mappinginventory.verifyHL7ToFHIRToHL7Mapping
+import org.junit.jupiter.api.Test
+
+class SPMtoSpecimen {
```
[the remark this answers is not in the document]
Comment (brick-green) [reply prime-router/src/testIntegration/kotlin/datatests/mappinginventory/spm/SPMtoSpecimen.kt:6]: I think that should be tested at the ORU_R01 level

**Pull request merged**